centraldogma/src/main/java/com/linecorp/decaton/centraldogma/CentralDogmaPropertySupplier.java

```diff
@@ -118,32 +126,43 @@ public <T> Optional<Property<T>> getProperty(PropertyDefinition<T> definition) {
             return Optional.empty();
         }
 
-        DynamicProperty<T> prop = new DynamicProperty<>(definition);
-        Watcher<JsonNode> child = rootWatcher.newChild(jsonNode -> jsonNode.path(definition.name()));
-        child.watch(node -> {
+        // note: cache DynamicProperties to avoid using too many child watchers if getProperty is called repeatedly.
+        // for most use cases though, this cache is only filled/read once.
+        return cachedProperties.computeIfAbsent(definition.name(), name -> {
+            DynamicProperty<T> prop = new DynamicProperty<>(definition);
+            Watcher<JsonNode> child = rootWatcher.newChild(jsonNode -> jsonNode.path(definition.name()));
+            child.watch(node -> {
+                try {
+                    setValue(prop, node);
+                } catch (Exception e) {
+                    // Catching Exception instead of RuntimeException, since
+                    // Kotlin-implemented DynamicProperty would throw checked exceptions
+                    logger.warn("Failed to set value updated from CentralDogma for {}", definition.name(), e);
+                }
+            });
             try {
+                JsonNode node = child.initialValueFuture().join().value(); //doesn't fail since it's a child watcher
                 setValue(prop, node);
             } catch (Exception e) {
                 // Catching Exception instead of RuntimeException, since
                 // Kotlin-implemented DynamicProperty would throw checked exceptions
-                logger.warn("Failed to set value updated from CentralDogma for {}", definition.name(), e);
+                logger.warn("Failed to set initial value from CentralDogma for {}", definition.name(), e);
             }
-        });
-        try {
-            JsonNode node = child.initialValueFuture().join().value(); //doesn't fail since it's a child watcher
-            setValue(prop, node);
-        } catch (Exception e) {
-            // Catching Exception instead of RuntimeException, since
-            // Kotlin-implemented DynamicProperty would throw checked exceptions
-            logger.warn("Failed to set initial value from CentralDogma for {}", definition.name(), e);
-        }
 
-        return Optional.of(prop);
+            return Optional.of(prop);
```

Comment:
Why not storing `DynamicProperty<?>` directly instead of wrapping with `Optional` ?

Reply:
Ah, no good reason. Fixed